keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/constraints/Ec2CanaryConstraintDeployHandler.kt

```diff
@@ -0,0 +1,166 @@
+import com.netflix.spinnaker.keel.model.toEchoNotification
+import com.netflix.spinnaker.keel.plugin.TaskLauncher
+import com.netflix.spinnaker.keel.retrofit.isNotFound
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import retrofit2.HttpException
+
+class Ec2CanaryConstraintDeployHandler(
+  private val taskLauncher: TaskLauncher,
+  private val cloudDriverService: CloudDriverService,
+  private val cloudDriverCache: CloudDriverCache,
+  private val imageService: ImageService
+) : CanaryConstraintDeployHandler {
+  companion object {
+    const val DEFAULT_KAYENTA_STORAGE_ACCOUNT = "s3-objects"
+    private val log by lazy { LoggerFactory.getLogger(Ec2CanaryConstraintDeployHandler::class.java) }
+  }
+
+  // TODO: The rest of Spinnaker should be refactored to refer to ec2 as ec2 instead of aws.
+  //  But for now, `CanaryConstraint.source` points to a source cluster (including cloud
+  //  provider) as Orca and Clouddriver would - that is, with "aws" as the provider when ec2
+  //  is intended.
+  override val supportedClouds = setOf("ec2", "aws")
+
+  override fun deployCanary(
+    constraint: CanaryConstraint,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment,
+    regions: Set<String>
+  ): Map<String, Task> {
+
+    val judge = "canary:${deliveryConfig.application}:${targetEnvironment.name}:${constraint.canaryConfigId}"
+
+    val image = runBlocking {
+      imageService.getLatestNamedImageWithAllRegionsForAppVersion(
+        appVersion = AppVersion.parseName(version.replace("~", "_")),
+        account = "test", // TODO: update when we have "default image account" configuration
+        regions = constraint.regions.toList())
+    }
+      ?.imageName ?: error("Image not found for $version in all requested regions ($regions)")
+
+    val regionalJobs = runBlocking {
+      val source = getSourceServerGroups(deliveryConfig.application, constraint)
+      require(regions.all { source.containsKey(it) }) {
+        "Source cluster ${constraint.source.cluster} not available in all canary regions"
+      }
+
+      regions.associateWith { region ->
+        val sourceServerGroup = source.getValue(region)
+        constraint.toStageBase(
+          cloudDriverCache = cloudDriverCache,
+          storageAccount = constraint.storageAccount ?: DEFAULT_KAYENTA_STORAGE_ACCOUNT,
+          app = deliveryConfig.application,
+          control = sourceServerGroup.toKayentaStageServerGroup(constraint.capacity, "baseline", image),
+          experiment = sourceServerGroup.toKayentaStageServerGroup(constraint.capacity, "canary", image)
+        )
+      }
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    return regionalJobs.map { (region, task) ->
+      val description = "Canary $version for ${deliveryConfig.application}/environment " +
+        "${targetEnvironment.name} in $region"
+
+      region to
```

Comment:
is this the start of returning a map? `region to` is such weird syntax, it would be a nice visual cue to me if you added a `return` even though it's not required by the compiler.

Reply:
Yeah, `collection.map { foo to bar }` returns a List<Pair> which is what `toMap` operates on. Where are you suggesting adding a return? `collection.map { return foo to bar }` doesn't appear to be valid. I can also see if refactoring around `regionalJobs.keys.associateWith { }` is more readable.

Reply:
Discussed offline - this syntax really through me for a loop, totally missed that this was inside the `return` & `.map {`. I'd prefer if the `try {` was extracted to a variable (`task`) and then this would become `region to task`. Fine either way though.